### PR TITLE
ZQS-1335: Implement Preparation of |COEF_int>

### PR DESCRIPTION
**Pull request by olgOk**
## Description

* Implementations of the PREP_int and PREP_int_prime

* Added basic tests for 'get_prep_int()'

*  Tests split

* Addressed @max-radin comments from[ ZQS-1336: Inverse of Block Encoding](https://github.com/zapatacomputing/benchq/pull/109)

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

**Comment by olgOk** [comment]
Tests look weird :)

File: tests/benchq/algorithms/test_lde_solver.py
Comment (AthenaCaesura) [inline tests/benchq/algorithms/test_lde_solver.py:1]: We should probably split this test up from one into two. They could also benefit from using @pytest.parameterize to test multiple inputs.

**Review by AthenaCaesura — CHANGES_REQUESTED** [review]

File: tests/benchq/algorithms/test_lde_solver.py
Comment (olgOk) [reply tests/benchq/algorithms/test_lde_solver.py:1]: Do you mean to split the tests up on basis cases (num qubits, circuit type, etc) and the amplitude testing?

File: src/benchq/algorithms/lde_solver.py [hunk 1]
```diff
@@ -194,7 +201,7 @@ def inverse_blockencoding(
 
     prep, prep_prime = control_prep(k=k, beta=beta)
     prep_prime_dag = prep_prime.inverse()
-    grid_qubits = ceil(np.log2(k)) + 1
+    grid_qubits = ceil(np.log2(k))
```
Comment (AthenaCaesura) [inline src/benchq/algorithms/lde_solver.py:204]: Didn't there used to be a +1 on this register? What happened?

File: tests/benchq/algorithms/test_lde_solver.py
Comment (AthenaCaesura) [reply tests/benchq/algorithms/test_lde_solver.py:1]: Yup! That works these tests look good!

**Review by AthenaCaesura — CHANGES_REQUESTED** [review]

File: src/benchq/algorithms/lde_solver.py (hunk 1 above)
Comment (olgOk) [reply src/benchq/algorithms/lde_solver.py:204]: Yes, it was! I think if there is +1, then the control qubit is outside of the PREP_int block

**Review by AthenaCaesura — APPROVED** [review]

File: tests/benchq/algorithms/test_lde_solver.py [hunk 2]
```diff
@@ -24,3 +37,105 @@
...
+            [1.0 * sympy.exp(-sympy.I * theta / 2), 0, 0, 0],
+            [0, 1.0 * sympy.exp(sympy.I * theta / 2), 0, 0],
+            [0, 0, 1.0 * sympy.exp(sympy.I * theta / 2), 0],
+            [0, 0, 0, 1.0 * sympy.exp(-sympy.I * theta / 2)],
+        ]
+    )
+    matrix_norm = float(custom_matrix.norm("fro"))
+    ZZ = CustomGateDefinition(
+        gate_name="ZZ", matrix=custom_matrix, params_ordering=(theta,)
+    )
```
[the remark this answers is not in the document]
Comment (olgOk) [reply tests/benchq/algorithms/test_lde_solver.py:110]: I think the idea in testing the inverse block encoding was to create a custom gate (any arbitrary but somewhat unique) and then count the number of total custom gates occurrences after the circuit construction by the gate attributes. @AthenaCaesura, do you think we should use  `orquestra.quantum.circuits.ZZ` or rather use a different name for this dummy custom gate, for example     `ZZ = CustomGateDefinition(gate_name="Dummy", ... )`? I simply took this example from Orquestra's tutorials, so if there is a better candidate for this dummy gate, we could change:

[the remark this answers is not in the document]
Comment (olgOk) [reply tests/benchq/algorithms/test_lde_solver.py]: Using proposed `orquestra.quantum.circuits.ZZ`

**Pull request merged**